service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,408 @@
+		}
+		batchCount := len(resp.Executions)
+		if batchCount <= 0 {
+			break
+		}
+
+		// send all tasks
+		for _, wf := range resp.Executions {
+			taskCh <- taskDetail{
+				execution: *wf.Execution,
+				attempts:  0,
+				hbd:       hbd,
+			}
+		}
+
+		succCount := 0
+		errCount := 0
+		// wait for counters indicate this batch is done
+	Loop:
+		for {
+			select {
+			case err := <-respCh:
+				if err == nil {
+					succCount++
+				} else {
+					errCount++
+				}
+				if succCount+errCount == batchCount {
+					break Loop
+				}
+			case <-ctx.Done():
+				return nil
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += succCount
+		hbd.ErrorCount += errCount
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return nil
+}
+
+func countWorkflowsWithRetry(ctx context.Context, svcClient workflowserviceclient.Interface, batchParams BatchParams) (*shared.CountWorkflowExecutionsResponse, error) {
+	var resp *shared.CountWorkflowExecutionsResponse
+	policy := backoff.NewExponentialRetryPolicy(rpcTimeout)
+	policy.SetMaximumInterval(batchParams.ActivityHeartBeatTimeout)
+	policy.SetExpirationInterval(batchParams.ActivityHeartBeatTimeout)
+
+	err := backoff.Retry(func() error {
+		var err error
+		resp, err = svcClient.CountWorkflowExecutions(ctx, &shared.CountWorkflowExecutionsRequest{
+			Domain: common.StringPtr(batchParams.DomainName),
+			Query:  common.StringPtr(batchParams.Query),
+		})
+		if err != nil {
+			getActivityLogger(ctx).Error("Failed to countWorkflowsWithRetry for batch operation task", tag.Error(err))
+		}
+
+		return err
+	}, policy, func(err error) bool {
+		return true
+	})
+
+	return resp, err
+}
+func scanWorkflowsWithRetry(ctx context.Context, svcClient workflowserviceclient.Interface, pageToken []byte, batchParams BatchParams) (*shared.ListWorkflowExecutionsResponse, error) {
+	var resp *shared.ListWorkflowExecutionsResponse
+	policy := backoff.NewExponentialRetryPolicy(rpcTimeout)
+	policy.SetMaximumInterval(batchParams.ActivityHeartBeatTimeout)
+	policy.SetExpirationInterval(batchParams.ActivityHeartBeatTimeout)
+	err := backoff.Retry(func() error {
+		var err error
+		resp, err = svcClient.ScanWorkflowExecutions(ctx, &shared.ListWorkflowExecutionsRequest{
+			PageSize:      common.Int32Ptr(int32(pageSize)),
+			Domain:        common.StringPtr(batchParams.DomainName),
+			NextPageToken: pageToken,
+			Query:         common.StringPtr(batchParams.Query),
+		})
+		if err != nil {
+			getActivityLogger(ctx).Error("Failed to scanWorkflowsWithRetry for batch operation task", tag.Error(err))
+		}
+
+		return err
+	}, policy, func(err error) bool {
+		return true
+	})
+
+	return resp, err
+}
+
+func startTaskProcessor(
+	ctx context.Context,
+	batchParams BatchParams,
+	taskCh chan taskDetail,
+	respCh chan error,
+	limiter *rate.Limiter) {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task := <-taskCh:
+			if isDone(ctx) {
```

Comment:
I dont think we need this check because of the upper case.. If the context gets cancelled while processing the task the RPC calls etc, will return an error and this method will return..

Reply:
Actually it's better to check it. If both ctx is canceled and a task comes, but task got processed first, then it will try to consume ratelimiter with the canceled and then return errors. 
See https://godoc.org/golang.org/x/time/rate#Limiter.WaitN

Since go select doesn't guarantee ordering, this could run into a loop until select eventually go into ctx.Done(). 
And there is no harm to check it here.